src/utils/CurlUtils.h

```diff
@@ -0,0 +1,98 @@
+/*
+ *  Copyright (C) 2023 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#pragma once
+
+#ifdef INPUTSTREAM_TEST_BUILD
+#include "../test/KodiStubs.h"
+#else
+#include <kodi/AddonBase.h>
+//! @Todo: forward CFile?
+#include <kodi/Filesystem.h>
+#endif
+
+#include <map>
+#include <string>
+#include <string_view>
+
+namespace UTILS
+{
+namespace CURL
+{
+
+enum class ReadStatus
+{
+  IS_EOF, // The end-of-file is reached
+  CHUNK_READ,
+  ERROR,
+};
+
+class ATTR_DLL_LOCAL CUrl
+{
+public:
+ /*!
+  * \brief Create CUrl.
+  * \param url The url of the file to download
+  */
+  CUrl(std::string_view url);
+  ~CUrl();
+
+ /*!
+  * \brief Open the url.
+  * \param isChunked Set true to enable chucked downloads
+  * \param isMediaStream Set true if the download is a media stream (audio/video/subtitles)
+  * \return Return HTTP status code, or -1 for any internal error
+  */
+  int Open(bool isChunked, bool isMediaStream = false);
+
+  void AddHeader(std::string_view name, std::string_view value);
+  void AddHeaders(const std::map<std::string, std::string>& headers);
+
+ /*!
+  * \brief Get an header from the HTTP response.
+  * \param name The header name
+  * \return The header value, or empty if none
+  */
+  std::string GetResponseHeader(std::string_view name);
+
+ /*!
+  * \brief Download / read a chunk, is required to enable chunks on Open() method.
+  * \param buffer[OUT] Buffer where to write the chunk data
+  * \param bufferSize The buffer size
+  * \param bytesRead[OUT] The chunk size read
+  * \return The read status
+  */
+  ReadStatus ReadChunk(void* buffer, size_t bufferSize, size_t& bytesRead);
+
+ /*!
+  * \brief Get the download speed in byte/s. To be called at the end of download.
+  */
+  double GetDownloadSpeed();
+
+ /*!
+  * \brief Get the total byte read of the download (total of chunks size).
+  */
+  size_t GetTotalByteRead() { return static_cast<size_t>(m_bytesRead); }
```

Comment:
Why not just return ssize_t ?

Reply:
`ssize_t` its a type used by kodi API interface only to return -1 value in case of internal errors on callback,
there is no reason to propagate this type on the ISA source code,
its not useful and make only more complex the code for nothing

Reply:
now that i am thinking better change m_bytesRead to size_t
and move the cast to ReadChunk since this is the total bytes may increase a lot